README.md
# Data_to_DB
# Python PostgreSQL 개발환경

- **언어:** Python 3.11
- **데이터베이스:** POSTGRES 15.5
- **Pip 버전:** Psycopg2 2.9.9
- **에디터:** PyCharm


- Python 패키지 설치 :
  pip install -r requirements.txt

  
- 실행하는 방법 :
  <code> python3 main.py </code>


- config 파일 구성하기 :

   config.json 파일에 id, password, host, port, name, schema 포함


- 데이터  입력 방법 : 

    sample.py - def insert_data()의 data에 데이터 넣기


- db에 연결하기 :

   config=sample.load_congig()로 정보 불러와  
conn, cur, engine = sample.postgres_con(config)로 연결

  
- table 만들기 :

   main.py - sample.create_table(conn,cur)로 테이블 생성


- DB에 입력된 data 확인 :

   main.py - sample.show_data(engine) 으로 확인


- table 삭제하기 :

  main.py - sample.drop_table(name,cur)로 테이블 삭제


- TODO :

  docker - container 적용하기


docker_run/history.txt
    1  git init
    2  git remote add origin git@github.com:honeybLee/DataHandling.git
    3  git pull origin master
    4  cat ~/.ssh/id_rsa.pub
    5  ssh-keygen
    6  cat ~/.ssh/id_rsa.pub
    7  git remote add origin git@github.com:honeybLee/DataHandling.git
    8  git pull origin master
    9  sudo apt-get update
   10  sudo apt install gnome-terminal
   11  dir
   12  cd ..
   13  dir
   14  cd wonch/
   15  cd download
   16  cd downloads
   17  cls
   18  cd 다운로드/
   19  약
   20  dir
   21  sudo apt-get install ./docker-desktop-4.25.2-amd64.deb 
   22  dir
   23  sudo apt-get install ./docker-desktop-4.25.2-amd64.deb
   24  sudo apt remove docker-desktop
   25  cd ..
   26  sudo apt purge docker-desktop
   27  cat /etc/hosts
   28  vim /etc/hosts
   29  sudo apt install vim
   30  sudo apt-get install ca-certificates
   31  sudo apt-get install curl
   32  sudo apt-get install gnupg
   33  sudo apt-get install lsb-release
   34  sudo mkdir -p /etc/apt/keyrings
   35  dir
   36  curl -fsSL https://download.docker.com/linux/ubuntu/gpg 
   37  200~sudo gpg --dearmor -o /etc/apt/keyrings/docker.gpg
   38  sudo gpg --dearmor -o /etc/apt/keyrings/docker.gpg
   39  sudo apt-get install curl
   40  sudo apt-get install gnupg
   41  sudo apt-get install lsb-release
   42  sudo mkdir -p /etc/apt/keyrings
   43  curl -fsSL https://download.docker.com/linux/gpg
   44  curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo gpg --dearmor -o /etc/apt/keyrings/docker.gpg
   45  sudo chmod a+r /etc/apt/keyrings/docker.gpg
   46  [200~echo   "deb [arch=$(dpkg --print-architecture) signed-by=/etc/apt/keyrings/docker.gpg] https://download.docker.com/linux/ubuntu \
   47    $(. /etc/os-release && echo "$VERSION_CODENAME") stable" | echo   "deb [arch=$(dpkg --print-architecture) signed-by=/etc/apt/keyrings/docker.gpg] https://download.docker.com/linux/ubuntu \
   48    $(. /etc/os-release && echo "$VERSION_CODENAME") stable" | \
   49  echo   "deb [arch=$(dpkg --print-architecture) signed-by=/etc/apt/keyrings/docker.gpg] https://download.docker.com/linux/ubuntu \
   50    $(. /etc/os-release && echo "$VERSION_CODENAME") stable" |   sudo tee /etc/apt/sources.list.d/docker.list > /dev/null
   51  sudo apt-get update
   52  sudo apt-get install docker-ce docker-ce-cli containerd.io docker-buildx-plugin docker-compose-plugin
   53  sudo docker version
   54  dir
   55  cd ..
   56  dir
   57  cd ..
   58  cd etc
   59  dir
   60  cd ..
   61  dir
   62  cd root
   63  cd ~
   64  dir
   65  cd ..
   66  cd sur
   67  dir
   68  cd wonch
   69  dir
   70  clear
   71  apt-get install git
   72  sudo apt-get install git
   73  dir
   74  cd 다운로드/
   75  dir
   76  cd pycharm-community-2023.2.5/
   77  dir
   78  cd bin
   79  dir
   80  sh pycharm.sh 
   81  python3 -V
   82  dir
   83  pycharm.sh
   84  sh pycharm.sh 
   85  cd ..
   86  dir
   87  sh pycharm.sh 
   88  cd ..
   89  dir
   90  wget -O - https://dbeaver.io/debs/dbeaver.gpg.key | sudo apt-key add -
   91  sudo wget -O - https://dbeaver.io/debs/dbeaver.gpg.key | sudo apt-key add -
   92  wget -O - https://dbeaver.io/debs/dbeaver.gpg.key | sudo apt-key add -
   93  echo "deb https://dbeaver.io/debs/dbeaver-ce /" | sudo tee /etc/apt/sources.list.d/dbeaver.list
   94  sudo apt-get update && sudo apt-get install dbeaver-ce
   95  wget -qO - https://www.mongodb.org/static/pgp/server-5.0.asc | sudo apt-key add -
   96  lsb_release -dc
   97  wget -qO - https://www.mongodb.org/static/pgp/server-4.4.asc | sudo apt-key add -
   98  sudo add-apt-repository 'deb [arch=amd64] https://repo.mongodb.org/apt/ubuntu focal/mongodb-org/4.4 multiverse'
   99  sudo apt install mongodb-org
  100  sudo yum erase $(sudo rpm -qa | grep mongodb-org)
  101  apt remove mongodb-org*
  102  sudo apt remove mongodb-org*
  103  apt-get purge mongodb-org*
  104  sudo apt-get purge mongodb-org*
  105  sudo rm -r /var/log/mongodb
  106  sudo rm -r /var/lib/mongodb
  107  sudo rm -r /etc/mongodb.conf
  108  sudo apt list --installed|grep mongo
  109  sudo wget -qO - https://www.mongodb.org/static/pgp/server-4.4.asc | sudo apt-key add -
  110  sudo add-apt-repository 'deb [arch=amd64] https://repo.mongodb.org/apt/ubuntu focal/mongodb-org/4.4 multiverse'
  111  sudo apt install mongodb-org
  112  E: 문제를 바로잡을 수 없습니다. 망가진 고정 패키지가 있습니다.
  113  sudo apt-get install mongodb-org-mongos
  114  sudo apt-get install libssl1.1
  115  sudo apt-get install libssl1.1=1.1.0
  116  sudo apt-get install mongodb-org
  117  sudo apt-get install mongodb-org-shell
  118  sudo apt-get install libssl
  119  sudo apt-get install libssl1
  120  sudo apt-get install libssl1.1
  121  sudo apt-get --fix-broken install
  122  sudo apt-get clean
  123  sudo apt-get install -f
  124  sudo dpkg --configure -a
  125  sudo apt-get update
  126  sudo rm /var/lib/apt/lists/* -vf
  127  sudo apt-get update
  128  sudo apt-get clean
  129  sudo apt-get autoclean
  130  sudo apt-get autoremove
  131  sudo dpkg --configure -a
  132  sudo apt-get update
  133  sudo apt update
  134  sudo apt install dirmngr gnupg apt-transport-https ca-certificates software-properties-common
  135  sudo wget -qO - https://www.mongodb.org/static/pgp/server-4.4.asc | sudo apt-key add -
  136  sudo add-apt-repository 'deb [arch=amd64] https://repo.mongodb.org/apt/ubuntu focal/mongodb-org/4.4 multiverse'
  137  sudo apt install mongodb-org
  138  sudo wget -qO - https://www.mongodb.org/static/pgp/server-6.0.asc | sudo apt-key add -
  139  echo "deb [ arch=amd64,arm64 ] https://repo.mongodb.org/apt/ubuntu jammy/mongodb-org/6.0 multiverse" | sudo tee /etc/apt/sources.list.d/mongodb-org-6.0.list
  140  sudo apt-get update
  141  sudo apt-get install -y mongodb-org
  142  clear
  143  sudo systemctl enable --now mongodb
  144  sudo systemctl enable --now mongod
  145  mongod
  146  mongo --eval 'db.runCommand({connectionStatus:1})'
  147  sudo mongo --eval 'db.runCommand({connectionStatus:1})'
  148  mongo localhost:4707
  149  dir

db_handler.py
import psycopg2
import json
import pandas as pd
from sqlalchemy import create_engine
import data_handler


# config.json 파일에서 정보 불러오기
def load_config(filename='config.json'):
    try:
        with open(filename, 'r') as config_file:
            config = json.load(config_file)
        return config
    except FileNotFoundError:
        raise ValueError(f"구성 파일 '{filename}'을 찾을 수 없습니다.")


def postgres_con(config):   # config.json 에 저장된 db정보로 접속
    user = config.get('db_user')
    password = config.get('db_password')
    host = config.get('db_host')
    port = config.get('db_port')
    db = config.get('db_name')

    if not (user and password and host and port and db):
        raise ValueError("DB 연결 정보를 찾을 수 없습니다. 환경 변수를 설정하세요.")

    pg_conn = psycopg2.connect(host=host, dbname=db, user=user, password=password, port=port)
    pg_cur = pg_conn.cursor()
    pg_engine = create_engine(f"postgresql+psycopg2://{user}:{password}@{host}:{port}/{db}")
    return pg_conn, pg_cur, pg_engine


def postgres_close(pg_conn, pg_cur):
    pg_conn.commit()
    pg_conn.close()
    pg_cur.close()


def create_table(conn, cur):
    # 'mydata' 이름의 테이블이 존재하지 않는다면 (coordinates, text_data, numeric_data) column을 갖는 테이블을 생성하는 쿼리
    create_table_sql = """
    CREATE TABLE IF NOT EXISTS mydata (
        coordinates integer[],
        text_data text,
        numeric_data double precision
    );
    """
    cur.execute(create_table_sql)
    conn.commit()


def insert_data(conn, cur):
    data = data_handler.read_data(file_path='./data.json')

    # mydata db의 (coordinates, text_data, numeric_data)칼럼에 값 넣기
    insert_data_sql = """
            INSERT INTO mydata (coordinates, text_data, numeric_data)
            VALUES (%s, %s, %s);"""

    for item in data:
        cur.execute(insert_data_sql, (item['coordinates'], item['text_data'], item['numeric_data']))
    conn.commit()


def drop_table(name, cur):
    query = f"drop table if exists {name}"  # 'name'의 테이블이 존재한다면 삭제시키는 쿼리
    cur.execute(query)





data_handler.py
import psycopg2
import json
import pandas as pd
from pandas import json_normalize
from sqlalchemy import create_engine
import db_handler


def read_data(file_path):
    with open(file_path, 'r', encoding='UTF8') as file:
        data = json.load(file)
    return data


def show_data(pg_engine):
    query = "select * from mydata"      # mydata 의 모든 값을 가져오는 쿼리
    pd.set_option('display.max_columns', None)  # df의 모든 column을 보이게 해줌
    df = pd.read_sql_query(query, pg_engine)
    print(df)


# data to json
def data_to_json(pg_engine):
    query = "select * from mydata"      # mydata 의 모든 값을 가져오는 쿼리
    df = pd.read_sql_query(query, pg_engine)
    df_json = df.to_json()
    return df_json


def select_data(engine, query):
    # select query example)
    # select numeric_data from mydata where numeric_data > 0.5
    # -> Mydata 테이블에서 numeric_data 값이 0.5 이상인 값들을 가져오는 쿼리
    df = pd.read_sql_query(query, engine)
    print("select_data :", df)
    return df


# 결측치 row 값 삭제
def missing_data(pg_engine):
    query = "select * from mydata"      # mydata 의 모든 값을 가져오는 쿼리
    df = pd.read_sql_query(query, pg_engine)
    df = df.dropna(axis=0)              # axis=1은 결측치 열이 제거
    return df


#data 중복값 삭제
def drop_duplicates(pg_engine):
    query = "select * from mydata"
    pd.set_option('display.max_columns', None)
    df = pd.read_sql_query(query, pg_engine)
    df.drop_duplicates(subset='text_data')
    return df




sample.py
import psycopg2
import json
import pandas as pd
from sqlalchemy import create_engine


def load_config(filename='config.json'):
    try:
        with open(filename, 'r') as config_file:
            config = json.load(config_file)
        return config
    except FileNotFoundError:
        raise ValueError(f"구성 파일 '{filename}'을 찾을 수 없습니다.")


def postgres_con(config):
    user = config.get('db_user')
    password = config.get('db_password')
    host = config.get('db_host')
    port = config.get('db_port')
    db = config.get('db_name')

    if not (user and password and host and port and db):
        raise ValueError("DB 연결 정보를 찾을 수 없습니다. 환경 변수를 설정하세요.")

    pg_conn = psycopg2.connect(host=host, dbname=db, user=user, password=password, port=port)
    pg_cur = pg_conn.cursor()
    pg_engine = create_engine(f"postgresql+psycopg2://{user}:{password}@{host}:{port}/{db}")
    return pg_conn, pg_cur, pg_engine


def postgres_close(pg_conn, pg_cur):
    pg_conn.commit()
    pg_conn.close()
    pg_cur.close()


def create_table(conn, cur):
    create_table_sql = """
    CREATE TABLE IF NOT EXISTS sample_data (
        coordinates integer[],
        text_data text,
        numeric_data double precision
    );

    """
    cur.execute(create_table_sql)
    conn.commit()


def insert_data(conn, cur):
    data = [
        ([[24, 8], [110, 8], [110, 16], [24, 16]],
         '미로나스국T리! 괴갑대쏘다',
         0.0017230819330160733),
        ([[10, 16], [56, 16], [56, 24], [10, 24]],
         '(P못 피스a(끼기',
         0.006269590404310938),
        ([[91, 73], [143, 73], [143, 89], [91, 89]], '사갑럽니다', 0.1783492257449009),
        ([[91, 91], [223, 91], [223, 105], [91, 105]],
         '바람이념신물더러온 그순간부터 ',
         0.10488713129482491),
        ([[170, 290], [200, 290], [200, 296], [170, 296]],
         '거파y3 너',
         0.017439958583245804),
        ([[99, 297], [167, 297], [167, 311], [99, 311]],
         '미아지키하아오',
         0.6818672154523849),
        ([[43, 313], [251, 313], [251, 375], [43, 375]],
         '바람이 붙다',
         0.6794710701367007),
        ([[65, 391], [223, 391], [223, 405], [65, 405]],
         '9리s임 가슴-속데간되다 그리움음 만니세인',
         0.012198750988838075)
    ]

    insert_data_sql = """
            INSERT INTO sample_data (coordinates, text_data, numeric_data)
            VALUES (%s, %s, %s);"""

    for item in data:
        cur.execute(insert_data_sql, (item[0], item[1], item[2]))
    conn.commit()


def drop_table(name, cur):
    query = f"drop table if exists {name}"
    cur.execute(query)


def show_data(pg_engine):
    query = "select * from sample_data"
    pd.set_option('display.max_columns', None)
    df = pd.read_sql_query(query, pg_engine)
    print(df)
